### PR TITLE
Rework assessment mutations

### PR DESCRIPTION
**Pull request by gigxz**
1. Restructure assessment API (see below)
2. Add `hud_values` column to AssessmentDetails table, to store JSON hud_values that are provided by the frontend when a form is submitted.

---

The previous api structure for saving assessments leads to a lot of duplication on the frontend. The newer structure more closely aligns to what the frontend is doing, and it keeps the Submit/extraction code in one place.

Old structure:
* `createAssessment(enrollmentId, inProgress: T|F, ...)` -- creates an assessment, and submits it if it's not WIP
* `saveAssessment(assessmentId, inProgress: T|F, ...)` -- saves an existing assessment, and submits it if it's not WIP

New structure:
* `saveAssessment(enrollmentId, assessmentId, ...)` -- saves a WIP assessment. creates a new one if it doesn't exist already (assessment ID is optional)
* `submitAssessment(enrollmentId, assessmentId, ...)` -- submits an assessment. creates a new one if it doesn't exist already (assessment ID is optional).

### CODE REVIEW

**Comment by gigxz** [comment]
@awisema @eanders let me know if you want to discuss any of this!

`hud_values` addition is based off convo with sandy this morning

File: drivers/hmis/app/graphql/mutations/submit_assessment.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      # Update values
+      assessment.assessment_detail.assign_attributes(values: values, hud_values: hud_values)
+      assessment.assign_attributes(
+        user_id: hmis_user.user_id,
+        date_updated: DateTime.current,
+        assessment_date: assessment_date ? Date.strptime(assessment_date) : assessment.assessment_date,
+      )
+
+      # TODO: validate hud_values, return any errors
+      # TODO: extract records from hud_values
```
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/submit_assessment.rb:51]: ⬆️

File: drivers/hmis/app/graphql/mutations/save_assessment.rb [hunk 2]
```diff
@@ -1,36 +1,54 @@
 module Mutations
   class SaveAssessment < BaseMutation
-    argument :assessment_id, ID, required: true
-    argument :values, Types::JsonObject, required: true
-    argument :in_progress, Boolean, required: false
+    description 'Create/Save assessment as work-in-progress'
+
+    argument :assessment_id, ID, 'Required if updating an existing assessment', required: false
+    argument :enrollment_id, ID, 'Required if saving a new assessment', required: false
+    argument :form_definition_id, ID, 'Required if saving a new assessment', required: false
```
Comment (eanders) [inline drivers/hmis/app/graphql/mutations/save_assessment.rb:7]: Not sure this is possible, but it'd be nice to have something like:
```suggestion
    argument :assessment_id, ID, 'Required if updating an existing assessment', required: persisted?
    argument :enrollment_id, ID, 'Required if saving a new assessment', required: !persisted?
    argument :form_definition_id, ID, 'Required if saving a new assessment', required: !persisted?
```

Comment (eanders) [reply drivers/hmis/app/graphql/mutations/save_assessment.rb:7]: I've seen something like that with a lambda, but not sure that's the right approach for a mutation.

**Review by eanders — APPROVED** [review]

File: drivers/hmis/app/graphql/mutations/save_assessment.rb (hunk 2 above)
Comment (gigxz) [reply drivers/hmis/app/graphql/mutations/save_assessment.rb:7]: We could put them in separate types and use ruby `oneOf` validation https://graphql-ruby.org/type_definitions/input_objects#oneof but I don't know if that's worth it. At the end of the day it is still a runtime check, none of the fields are required in the schema.

**Comment by gigxz** [comment]
Can I go ahead and merge this one?

**Comment by eanders** [comment]
Sure, I was just waiting on Sandy's review.

**Comment by gigxz** [comment]
Ok. I'm going to merge for the sake of several dependent PRs/branches, and @awisema can make any changes to the SubmitAssessment mutation as needed/desired once he gets in there 🙂

**Pull request merged**